scipy/spatial/transform/tests/test_rotation.py

```diff
@@ -658,39 +652,144 @@ def test_as_euler_degenerate_symmetric_axes():
         [15, 0, 60],
         [35, 0, 75],
         [60, 180, 35],
-        [15, -180, 25],
-        ])
+        [15, -180, 25]])
 
     with pytest.warns(UserWarning, match="Gimbal lock"):
-        for axis1 in ['x', 'y', 'z']:
-            for axis2 in ['x', 'y', 'z']:
-                if axis1 == axis2:
-                    continue
+        for seq_tuple in permutations('xyz'):
+            # Extrinsic rotations
+            seq = ''.join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
+            rotation = Rotation.from_euler(seq, angles, degrees=True)
+            mat_expected = rotation.as_matrix()
 
-                # Extrinsic rotations
-                seq = axis1 + axis2 + axis1
-                rotation = Rotation.from_euler(seq, angles, degrees=True)
-                mat_expected = rotation.as_matrix()
+            angle_estimates = rotation.as_euler(seq, degrees=True)
+            mat_estimated = Rotation.from_euler(
+                seq, angle_estimates, degrees=True
+                ).as_matrix()
+
+            assert_array_almost_equal(mat_expected, mat_estimated)
+
+            # Intrinsic rotations
+            seq = seq.upper()
+            rotation = Rotation.from_euler(seq, angles, degrees=True)
+            mat_expected = rotation.as_matrix()
+
+            angle_estimates = rotation.as_euler(seq, degrees=True)
+            mat_estimated = Rotation.from_euler(
+                seq, angle_estimates, degrees=True
+                ).as_matrix()
+
+            assert_array_almost_equal(mat_expected, mat_estimated)
 
-                angle_estimates = rotation.as_euler(seq, degrees=True)
-                mat_estimated = Rotation.from_euler(
-                    seq, angle_estimates, degrees=True
-                    ).as_matrix()
 
-                assert_array_almost_equal(mat_expected, mat_estimated)
+def test_as_euler_compare_algorithms():
+    rnd = np.random.RandomState(0)
+    n = 10000
+    angles = np.empty((n, 3))
+    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
+    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
 
-                # Intrinsic rotations
-                seq = seq.upper()
-                rotation = Rotation.from_euler(seq, angles, degrees=True)
-                mat_expected = rotation.as_matrix()
+    # asymmetric axes
+    angles[:, 1] = rnd.uniform(low=-np.pi / 2, high=np.pi / 2, size=(n,))
+    for seq_tuple in permutations('xyz'):
+        # Extrinsic rotations
+        seq = ''.join(seq_tuple)
+        rot = Rotation.from_euler(seq, angles)
+        assert_allclose(rot._as_euler_from_matrix(seq), rot.as_euler(seq),
+                        atol=0, rtol=1e-11)
```

Comment:
`rtol=1e-11` is not very close to numerical precision for double, which is about `1e-16 - 1e-15`.

In this case I think it makes sense to investigate which method is more accurate by comparison to ground truth angles.

Again why do we even need to compare between methods instead of simply to the ground truth (for this particular case with random non-degenerate angles)?

Reply:
Interesting, maybe it is accumulated errors with the trigonometric functions?  I tried all possible tests: both methods against one another, and ground truth to any of the algorithms. They all give similar results:

Comparing `_as_euler_from_matrix` to `as_euler`:
```
E   AssertionError: 
E   Not equal to tolerance rtol=1e-12, atol=0
E   
E   Mismatched elements: 2 / 30000 (0.00667%)
E   Max absolute difference: 3.17523785e-13
E   Max relative difference: 1.99354194e-12
E    x: array([[ 0.306704, -0.338749,  1.559914],
E          [ 1.352075, -1.441499, -2.009346],
E          [ 0.645681,  1.329838, -0.697288],...
E    y: array([[ 0.306704, -0.338749,  1.559914],
E          [ 1.352075, -1.441499, -2.009346],
E          [ 0.645681,  1.329838, -0.697288],...
```

Ground truth to `as_euler`:
```
E   AssertionError: 
E   Not equal to tolerance rtol=1e-12, atol=0
E   
E   Mismatched elements: 1 / 30000 (0.00333%)
E   Max absolute difference: 3.17523785e-13
E   Max relative difference: 1.42665666e-12
E    x: array([[ 0.306704, -0.338749,  1.559914],
E          [ 1.352075, -1.441499, -2.009346],
E          [ 0.645681,  1.329838, -0.697288],...
E    y: array([[ 0.306704, -0.338749,  1.559914],
E          [ 1.352075, -1.441499, -2.009346],
E          [ 0.645681,  1.329838, -0.697288],...
```
Comparing ground truth `_as_euler_from_matrix`:
```
E   AssertionError: 
E   Not equal to tolerance rtol=1e-12, atol=0
E   
E   Mismatched elements: 1 / 30000 (0.00333%)
E   Max absolute difference: 3.94795308e-13
E   Max relative difference: 1.99354194e-12
E    x: array([[ 0.306704, -0.338749,  1.559914],
E          [ 1.352075, -1.441499, -2.009346],
E          [ 0.645681,  1.329838, -0.697288],...
E    y: array([[ 0.306704, -0.338749,  1.559914],
E          [ 1.352075, -1.441499, -2.009346],
E          [ 0.645681,  1.329838, -0.697288],...
```
So whatever causes this difference, the difference is there independent of the algorithm used.

And to answer your last question: 
To be honest, I didn't really think we had to do this comparison, but I imagined it would be better to leave it here anyways so other maintainers can be more assured there won't be any differences.

Reply:
I see, interesting. So both methods have similar performance relative to the ground truth and between each other. 

How about this then: modify this test such that both solutions are compared against the ground truth with the same `rtol`. It can be interpreted as "both methods have the same accuracy". And reducing `rtol` further will cause the error.

Can it be done the same way in the test with special cases?

> To be honest, I didn't really think we had to do this comparison, but I imagined it would be better to leave it here anyways so other maintainers can be more assured there won't be any differences.

I guess I was sloppy at some point of the conversation. I thought you've found some other reasons to test methods against each other.

Anyway, I think my last suggestion above can justify keeping them both.

Reply:
Also what is the error statistics like mean value and standard deviation? We can compare such averaged characteristics too in a test.

Reply:
> How about this then: modify this test such that both solutions are compared against the ground truth with the same `rtol`. It can be interpreted as "both methods have the same accuracy". And reducing `rtol` further will cause the error.

What if (for non degenerate cases) I add lines to the two original `_test_as_euler_asymmetric_axes` and `_test_as_euler_symmetric_axes` then? Since it is basically the same test, this will probably be a lot clearer to read (two new lines in each test instead of a whole new test.)

> Can it be done the same way in the test with special cases?

The problem with the special cases (testing directly against the ground truth) is that the tests would be very awkward to implement and debug, and prone to error (since I'd have to calculate the sum and difference between `angles[0]` and `angles[2]`, but the signs should change according to the parity and symmetry...)
This is the reason why I'm testing one algorithm to the other: they will give (up to some numerical error) the same values (since they the decision process of which angle to assign `zero` to in the degenerate cases is the same in both algorithms.)

> Also what is the error statistics like mean value and standard deviation? We can compare such averaged characteristics too in a test.

Sadly I'm not very familiar with pytest... How would I proceed doing this?

Reply:
For non degenerate cases, I propose to remove `test_as_euler_compare_algorithms` completely and change the original functions to this:
```
def test_as_euler_asymmetric_axes():
    rnd = np.random.RandomState(0)
    n = 1000
    angles = np.empty((n, 3))
    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
    angles[:, 1] = rnd.uniform(low=-np.pi / 2, high=np.pi / 2, size=(n,))
    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))

    for seq_tuple in permutations('xyz'):
        # Extrinsic rotations
        seq = ''.join(seq_tuple)
        rotation = Rotation.from_euler(seq, angles)
        assert_allclose(angles, rotation.as_euler(seq), atol=0, rtol=1e-12)
        assert_allclose(angles, rotation._as_euler_from_matrix(seq),
                        atol=0, rtol=1e-12)
        # Intrinsic rotations
        seq = seq.upper()
        rotation = Rotation.from_euler(seq, angles)
        assert_allclose(angles, rotation.as_euler(seq), atol=0, rtol=1e-12)
        assert_allclose(angles, rotation._as_euler_from_matrix(seq),
                        atol=0, rtol=1e-12)


def test_as_euler_symmetric_axes():
    rnd = np.random.RandomState(0)
    n = 1000
    angles = np.empty((n, 3))
    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
    angles[:, 1] = rnd.uniform(low=0, high=np.pi, size=(n,))
    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))

    for seq_tuple in permutations('xyz'):
        # Extrinsic rotations
        seq = ''.join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
        rotation = Rotation.from_euler(seq, angles)
        assert_allclose(angles, rotation.as_euler(seq), atol=0, rtol=1e-13)
        assert_allclose(angles, rotation._as_euler_from_matrix(seq),
                        atol=0, rtol=1e-9) 
        # Intrinsic rotations
        seq = seq.upper()
        rotation = Rotation.from_euler(seq, angles)
        assert_allclose(angles, rotation.as_euler(seq), atol=0, rtol=1e-13)
        assert_allclose(angles, rotation._as_euler_from_matrix(seq),
                        atol=0, rtol=1e-9)
```
Not only it is a lot cleaner, like I said before, but it also shows clearly that my method passes for smaller values of `rtol` (that I just checked).

Reply:
For the degenerate cases, I'd argue though that the actual important test is the one `test_as_euler_degenerate_symmetric_axes`, so if we want to test for numerical accuracy, we should probably test this one (even though I'm not completely sure why we are testing the rotation matrices instead of the quaternions...)

The importance of `test_as_euler_degenerate_compare_algorithms`, for me, is to show that in degenerate cases, both algorithms are taking the same **decisions** and giving the "same" values, meaning that we are deciding to assign `zero` to the same angle in all cases (any sequence, extrinsic, intrinsic, etc). 

This way, anyone already using this function on their codes won't suddenly have their scripts giving out different angles (even though they can represent equivalent rotations).

Reply:
> Sadly I'm not very familiar with pytest... How would I proceed doing this?

Not related to pytest. Instead of comparing individual angles we can computed `np.mean` and `np.std` for angle differences. It should give a meaningful picture of expected errors for both methods. 

You can add 4 more assertions for these statistic values.

> For non degenerate cases, I propose to remove test_as_euler_compare_algorithms completely and change the original functions to this:

Looks good.

> The importance of test_as_euler_degenerate_compare_algorithms, for me, is to show that in degenerate cases, both algorithms are taking the same decisions and giving the "same" values, meaning that we are deciding to assign zero to the same angle in all cases (any sequence, extrinsic, intrinsic, etc).

> This way, anyone already using this function on their codes won't suddenly have their scripts giving out different angles (even though they can represent equivalent rotations).

Totally makes sense to me.

But besides that, do we also want to quantitatively assess whether the new method is more or less accurate than the previous for degenerate cases?

Reply:
Just deleted `test_as_euler_compare_algorithms` and integrated the non degenerate tests in the other functions, as I described! :)

> But besides that, do we also want to quantitatively assess whether the new method is more or less accurate than the previous for degenerate cases?

Very good question, I'm not really sure about the answer! Since we are comparing the actual rotation matrices instead of the angles, we are applying inverse trigonometric functions and then re-applying trigonometric functions again... I'm not sure how to best compare those in a significant way tbh!

> Not related to pytest. Instead of comparing individual angles we can compute np.mean and np.std for angle differences. It should give a meaningful picture of expected errors for both methods.
> 
> You can add 4 more assertions for these statistic values.
> 
> Or perhaps not mean and std but rather RMS = hypot(mean, std). But mean is also of interest, i. e. do algorithms are biased somehow?

Do you mean something like the following code?
> assert_allclose(np.std(angles - rotation.as_euler(seq)), np.std(angles - rotation._as_euler_from_matrix(seq)))

Or something else entirely? I haven't seen any other test in `test_rotation` implementing anything like this using `np.std` or any kind of mean... I'm not sure how to implement any test that is significant (statistics are definitely **not** something I have a lot of experience with, I'm afraid).

Reply:
This is my first time ever contributing to something like this btw, so I'm sorry if I'm not very comfortable taking decisions!

Reply:
> Very good question, I'm not really sure about the answer! Since we are comparing the actual rotation matrices instead of the angles, we are applying inverse trigonometric functions and then re-applying trigonometric functions again... I'm not sure how to best compare those in a significant way tbh!

In `test_as_euler_degenerate_compare_algorithms` you do compare accuracies already. But there are very few angles, so it doesn't make sense to compute any statistics there. 

Looks like nothing actually blocks us by generating large amount of random tests similar to that: assign the second angle fixed (or random in the narrow segment near the degenerate angle) and generate the first and the second angles randomly.

> Do you mean something like the following code?

Something quite similar. I had in my mind something like that:

```
error = rotation.as_euler(seq) - angles
mean = np.mean(error, axis=0)
std = np.mean(error, axis=0)
rms = np.hypot(mean, rms)
assert np.all(np.abs(mean)) < ...
assert np.all(rms) < ...

# Do the same with rotation._as_euler_from_matrix(seq)
```

> I'm not sure how to implement any test that is significant (statistics are definitely not something I have a lot of experience with, I'm afraid).

I think basic engineering judgment is sufficient here. If we take say 1000 rotations then the error distribution must be somewhat representative and checking mean, std, rms looks reasonable.

Reply:
Do you mean this for the non degenerate cases or the degenerate cases?

The problem with `error = rotation.as_euler(seq) - angles` is that, for degenerate cases, we come back to that problem I was talking about: `rotation.as_euler(seq)` won't give out the same angles as `angles`, we can only assume a either a sum or diff between them, and the exact expression of this sum and diff in particular will be very hard to implement in a way we can be **sure** it is correct, because of the all of the cases (parity, extrinsic, etc) and I find that if I did an implementation of this I wouldn't even know how to verify its correctness.

Reply:
> The problem with error = rotation.as_euler(seq) - angles is that, for degenerate cases, we come back to that problem I was talking about: rotation.as_euler(seq) won't give out the same angles as angles, we can only assume a either a sum or diff between them, and the exact expression of this sum and diff in particular will be very hard to implement in a way we can be sure it is correct, because of the all of the cases (parity, extrinsic, etc) and I find that if I did an implementation of this I wouldn't even know how to verify its correctness.

OK, I see. Let's not bother with them.

> If you mean the non-degenerate cases:

> Isn't the use of the assert methods of pytest instead of standard deviation and mean of the error already a kind of design decision for the test files?

I think we should have a flexible mindset in scipy. 

Here `test_as_euler_asymmetric_axes` and `test_as_euler_symmetric_axes` became more of an accuracy benchmark rather than a mere test. And we do need benchmarks for this kind of algorithms.

Both type of assertions make sense. The current assertions effectively checks the worst case accuracy, and mean / rms will check an average / typical accuracy.

Reply:
Implemented `test_as_euler_benchmark_algorithms` as you proposed.

Everything passes with `rms_max = 1e-14`, below it doesn't pass anymore.

Everything passes with `std_max = 1e-15`, and for symmetric cases, `as_euler` can go to `std_max = 1e-16`, while `_as_euler_from_matrix` can't.

I think these are close to the float precision you talked about. Since we are close to it, maybe these exact values won't work if we try with more tests. 

But this should be enough to show that both methods give exact results.

Reply:
I see, looks good!

Only I would keep the original `assert_allclose`, they won't hurt. No reason to reduce number of tests in general.

Reply:
I kept the assert tests, they are in the original functions! These std and RMS tests are in a different test function I just created. Since we are testing different values I thought it was better to keep tests separated for easier debugging eventually.

Reply:
Then I suggest removing modifications in other tests, except maybe this --- `n = 500` (to strengthen the test).

Just to keep the diff at minimum.

Reply:
> Then I suggest removing modifications in other tests

This is what I had done before, but then we decided to do it like this because, modifying the functions only adds 4 lines of code, while moving separating then creates a whole other function and adds lots of lines.

Moreover, for symmetric axes, I greatly simplified this:
```
for axis1 in ['x', 'y', 'z']:
    for axis2 in ['x', 'y', 'z']:
        if axis1 == axis2:
            continue
        seq = axis1 + axis2 + axis1
```
To this:
```
for seq_tuple in permutations('xyz'):
    seq = ''.join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
```
In the original functions, reducing the depth of the idents and making it as compact as the tests for assymmetric axes.
The original for assymetric is this:
```
for seq_tuple in permutations('xyz'):
    seq = ''.join(seq_tuple)
```

This will affect the number of diffs anyways, but I think this is worth it.